static_frame/core/frame.py

```diff
@@ -5968,6 +6013,27 @@ def to_latex(self,
                 )
         return repr(self.display(config))
 
+    def to_msgpack(self) -> 'msgpack':
+        '''
+        Return a msgpack.
+        '''
+        import msgpack
+        import msgpack_numpy
+        def cast_msgpack(input):
+            try:
+                data = msgpack.packb(input, default=msgpack_numpy.encode) #try coercing standard and numpy datatypes
+            except TypeError:
+                #This inserts a magic character with any arbitrary datatype that msgpack_numpy doesn't support
```

Comment:
Specifically np.datetime64. which I covered in test_frame_from_msgpack_b, it would be beneficial to convert that specifically to an int and a custom datatype, however I was guessing there might be many cases like this. I also figured msgpack_numpy developers would have already added a clean way to do it if it was easy.

I added the hack to ensure we would get the same data types when running from_msgpack, however if the goal is to share the msgpack with other applications, using a plain string as the fallback would be more desirable.

Reply:
btw here is the example use case using yfinance that I built to play around with that led me to discovering np.datetime64 was unsupported:

[KungFu/Python/DataStructures/Array/StaticFrameFromSymbol.py](https://github.com/MadisonAster/KungFu/blob/master/Python/DataStructures/Arrays/StaticFrameFromSymbol.py)

[KungFu/Python/DataStructures/Binary/MessagePack.py](https://github.com/MadisonAster/KungFu/blob/master/Python/DataStructures/Binary/MessagePack.py)

Reply:
Interesting: so is it only for np.datetime64 types that we need the "Magic Character Hack"? If that is the case perhaps we can just identify that type and perform a special encoding/decoding. I think it is the case that all np.datetime64 can be represented in integers, and be converted back to np.datetime64 as long as you know the dtype. np.datetime64 could then be encoded as a pair of int array and string dtype representation.

```python
>>> a1 = np.array(['2010', '2013'], dtype=np.datetime64)
>>> a1
array(['2010', '2013'], dtype='datetime64[Y]')
>>> a1.astype(int)
array([40, 43])
>>> a1.dtype.str
'<M8[Y]'
>>> np.array(a1.astype(int), dtype=a1.dtype.str)
array(['2010', '2013'], dtype='datetime64[Y]')
```

Reply:
it applies to timedelta64, and datetime64. I pushed new commits last night that take care of the same thing with dtype, but still using the magic character to pass the information across. It's only working for 1D arrays like index and columns at the moment, and I need to get it working for column types as well. See test_frame_from_msgpack_d

Reply:
We can almost certainly pass the type information for this with a single character, instead of using the hack. I need to figure that part of msgpack out still.

Reply:
I see. Encoding `timedelta64` as a pair of int array, dtype string might also be possible.